utils/gcp/object_storage.cc

```diff
@@ -0,0 +1,820 @@
+
+            auto res = co_await _impl->send_with_retry(_session_path
+                , GCP_OBJECT_SCOPE_READ_WRITE
+                , std::make_pair([&](output_stream<char>&& os_in) -> future<> {
+                    auto os = std::move(os_in);
+                    for (auto& buf : bufs) {
+                        co_await os.write(buf.share());
+                    }
+                    co_await os.flush();
+                    co_await os.close();
+                }, len)
+                , ""s // no content type
+                , httpclient::method_type::PUT
+                , rest::key_values({ { CONTENT_RANGE, range } })
+            );
+
+            switch (res.result()) {
+            case status_type::ok:
+            case status_type::created:
+                _completed = true;
+                gcp_storage.debug("{}:{} completed ({} bytes)", _bucket, _object_name, offset+len);
+                co_return; // done and happy
+            default:
+                if (int(res.result()) == 308) {
+                    uint64_t first = 0, new_last = 0;
+                    if (parse_response_range(res.reply, first, new_last) && last != new_last) {
+                        auto written = (new_last + 1) - offset;
+
+                        gcp_storage.debug("{}:{} partial upload ({} bytes)", _bucket, _object_name, written);
+
+                        if (!final && (len - written) < min_gcp_storage_chunk_size) {
+                            written = len - std::min(min_gcp_storage_chunk_size, len);
+                        }
+
+                        auto to_remove = written;
+                        while (to_remove) {
+                            auto& buf = bufs.front();
+                            auto size = std::min(to_remove, buf.size());
+                            buf.trim_front(size);
+                            if (buf.empty()) {
+                                bufs.pop_front();
+                            }
+                            to_remove -= size;
+                        }
+                        offset += written;
+                        len -= written;
+                        auto total = std::accumulate(bufs.begin(), bufs.end(), size_t{}, [](size_t s, auto& buf) {
+                            return s + buf.size();
+                        });
+                        assert(len == total);
+                        retries = 0; // reset.
+                        continue;
+                    }
+                    // incomplete. ok for partial
+                    gcp_storage.debug("{}:{} chunk {}:{} done", _bucket, _object_name, offset, offset+len);
+                    co_return;
+                }
+                if (reply::classify_status(res.result()) == reply::status_class::server_error && ++retries < 4) {
+                    gcp_storage.debug("{}:{} chunk {}:{} server error. backoff {}...", _bucket, _object_name, offset, offset+len, retries);
+                    co_await backoff(retries);
```

Comment:
They mention exponential backoffs in their documentation

Reply:
Done.